controllers/space/space_controller.go

```diff
@@ -0,0 +1,345 @@
+	// if UserAccount ready condition is already set to false, then don't update if no message is provided
+	// also, don't update when NSTemplateSet ready condition is in an invalid state
+	if !found || readyCond.Status == corev1.ConditionUnknown || readyCond.Status == corev1.ConditionFalse {
+		logger.Info("NSTemplateSet is not ready", "ready-condition", readyCond)
+		if err := r.setStatusFromNSTemplateSet(space, readyCond.Reason, readyCond.Message); err != nil {
+			return false, err
+		}
+		return true, nil
+	}
+	return false, r.setStatusReady(space)
+}
+
+func (r *Reconciler) newNSTemplateSet(memberOperatorNS string, space *toolchainv1alpha1.Space) (*toolchainv1alpha1.NSTemplateSet, error) {
+	// look-up the NSTemplateTier
+	tmplTier := &toolchainv1alpha1.NSTemplateTier{}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: space.Namespace,
+		Name:      space.Spec.TierName,
+	}, tmplTier); err != nil {
+		return nil, err
+	}
+	// create the NSTemplateSet from the NSTemplateTier
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: memberOperatorNS,
+			Name:      space.Name,
+			Finalizers: []string{
+				toolchainv1alpha1.FinalizerName,
+			},
+		},
+		Spec: toolchainv1alpha1.NSTemplateSetSpec{
+			TierName: space.Spec.TierName,
+			ClusterResources: &toolchainv1alpha1.NSTemplateSetClusterResources{
+				TemplateRef: tmplTier.Spec.ClusterResources.TemplateRef,
+			},
+			Namespaces: make([]toolchainv1alpha1.NSTemplateSetNamespace, len(tmplTier.Spec.Namespaces)),
+		},
+	}
+	for i, tmpl := range tmplTier.Spec.Namespaces {
+		nsTmplSet.Spec.Namespaces[i] = toolchainv1alpha1.NSTemplateSetNamespace{
+			TemplateRef: tmpl.TemplateRef,
+		}
+	}
+
+	return nsTmplSet, nil
+}
+
+func (r *Reconciler) ensureSpaceDeletion(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	logger.Info("terminating Space")
+	if deleted, err := r.deleteNSTemplateSet(logger, space); err != nil {
+		logger.Error(err, "failed to delete the NSTemplateSet")
+		return false, r.setStatusTerminatingFailed(space, "failed to delete the NSTemplateSet")
+	} else if deleted {
+		if err := r.setStatusTerminating(space); err != nil {
+			logger.Error(err, "error updating status")
+			return false, err
+		}
+		return true, nil // requeue until NSTemplateSet is fully deleted
+	}
+	// Remove finalizer from Space
+	util.RemoveFinalizer(space, toolchainv1alpha1.FinalizerName)
+	if err := r.Client.Update(context.TODO(), space); err != nil {
+		logger.Error(err, "failed to remove finalizer")
+		return false, r.setStatusTerminatingFailed(space, "failed to remove finalizer")
+	}
+	logger.Info("removed finalizer")
+	// no need to update the status of the Space once the finalizer has been removed, since
+	// the resource will be deleted
+	return false, nil // no need to requeue
+}
+
+// deleteNSTemplateSet triggers the deletion of the NSTemplateSet on the target member cluster.
```

Comment:
May I request to add a clarification comment about what the bool return value represents? Looking at the code it's not clear to me whether it is indicating whether the NSTemplateSet is deleted or if it's supposed to trigger a requeue. To me the function looks to treat it as a requeue but where the function is called in the `ensureSpaceDeletion` func the variable is named `deleted` so I'm confused

Reply:
I clarified a bit the comment in 35a2d2f, but note that after d34a855, the controller does not requeue explicitly while the NSTemplateSet is being deleted (since the controller watches the NSTemplateSets on the member clusters, a new reconcile loop will automatically be triggered when the NSTemplateSet is deleted). 
Please let me know if it's better like this now ;)